datastore/transaction.go

```diff
@@ -159,47 +157,115 @@ func (c *Client) NewTransaction(ctx context.Context, opts ...TransactionOption)
 	return c.newTransaction(ctx, newTransactionSettings(opts))
 }
 
-func (c *Client) newTransaction(ctx context.Context, s *transactionSettings) (_ *Transaction, err error) {
-	req := &pb.BeginTransactionRequest{
-		ProjectId:  c.dataset,
-		DatabaseId: c.databaseID,
+func (t *Transaction) parseTransactionOptions() (*pb.TransactionOptions, string) {
+	if t.settings == nil {
+		return nil, ""
 	}
-	if s.readOnly {
-		ctx = trace.StartSpan(ctx, "cloud.google.com/go/datastore.Transaction.ReadOnlyTransaction")
-		defer func() { trace.EndSpan(ctx, err) }()
 
+	if t.settings.readOnly {
 		ro := &pb.TransactionOptions_ReadOnly{}
-		if !s.readTime.AsTime().IsZero() {
-			ro.ReadTime = s.readTime
+		if !t.settings.readTime.AsTime().IsZero() {
+			ro.ReadTime = t.settings.readTime
 		}
 
-		req.TransactionOptions = &pb.TransactionOptions{
+		return &pb.TransactionOptions{
 			Mode: &pb.TransactionOptions_ReadOnly_{ReadOnly: ro},
-		}
-
-	} else if s.prevID != nil {
-		ctx = trace.StartSpan(ctx, "cloud.google.com/go/datastore.Transaction.ReadWriteTransaction")
-		defer func() { trace.EndSpan(ctx, err) }()
+		}, "cloud.google.com/go/datastore.Transaction.ReadOnlyTransaction"
+	}
 
-		req.TransactionOptions = &pb.TransactionOptions{
+	if t.settings.prevID != nil {
+		return &pb.TransactionOptions{
 			Mode: &pb.TransactionOptions_ReadWrite_{ReadWrite: &pb.TransactionOptions_ReadWrite{
-				PreviousTransaction: s.prevID,
+				PreviousTransaction: t.settings.prevID,
 			}},
-		}
+		}, "cloud.google.com/go/datastore.Transaction.ReadWriteTransaction"
 	}
-	resp, err := c.client.BeginTransaction(ctx, req)
+	return nil, ""
+}
+
+// beginTransaction makes BeginTransaction rpc
+func (t *Transaction) beginTransaction() (txnID []byte, err error) {
+
+	req := &pb.BeginTransactionRequest{
+		ProjectId:  t.client.dataset,
+		DatabaseId: t.client.databaseID,
+	}
+
+	txOptionsPb, spanName := t.parseTransactionOptions()
+	if txOptionsPb != nil {
+		t.ctx = trace.StartSpan(t.ctx, spanName)
+		defer func() { trace.EndSpan(t.ctx, err) }()
+		req.TransactionOptions = txOptionsPb
+	}
+
+	resp, err := t.client.client.BeginTransaction(t.ctx, req)
 	if err != nil {
 		return nil, err
 	}
-	return &Transaction{
-		id:        resp.Transaction,
+	return resp.Transaction, nil
+}
+
+// beginLaterTransaction makes BeginTransaction rpc if transaction has not yet started
+func (t *Transaction) beginLaterTransaction() (err error) {
+	if t.state != transactionStateNotStarted {
+		return nil
+	}
+
+	// Obtain state lock since the state needs to be updated
+	// after transaction has started
+	t.stateLock.Lock()
+	defer t.stateLock.Unlock()
+	if t.state != transactionStateNotStarted {
+		return nil
+	}
+
+	txnID, err := t.beginTransaction()
+	if err != nil {
+		return err
+	}
+
+	t.startProgress(txnID)
+	return nil
+}
+
+// Acquire state lock if transaction has not started
```

Comment:
Could you add to the comment for `acquireLock` more about what it is, what its return value signifies, and how it's used in other parts of the PR? It does a lot more than simply acquiring the state lock.

Also because this function does a lot more than just acquiring the state lock, could you rename the function to reflect that?

Reply:
Resolved in https://github.com/googleapis/google-cloud-go/pull/10063